### PR TITLE
add additional interpolation modes for torch.quantile

### LINKED ISSUE #48523: Implement additional quantile modes
This issue is to track the request for additional quantile modes: lower, higher, nearest, and midpoint. These modes are available in NumPy: https://numpy.org/doc/stable/reference/generated/numpy.quantile.html?highlight=quantile#numpy.quantile and requested by @matthewfeickert here https://github.com/pytorch/pytorch/issues/38349#issuecomment-717606900.

cc @mruberry @rgommers @heitorschueroff

### PR DESCRIPTION
**Pull request by RockingJavaBean**
Fix #48523
Related  #38349

**BC-breaking Note:**

This PR updates PyTorch's quantile function to add additional interpolation methods `lower`, `higher`, `nearest`, and `midpoint`, and these interpolation methods are currently supported by NumPy.

New parameter `interpolation` is added to the signature for both `torch.quantile` and `torch.nanquantile` functions.

- `quantile(input, q, dim=None, interpolation='linear', keepdim=False, *, out=None) -> Tensor`
- `nanquantile(input, q, dim=None, interpolation='linear', keepdim=False, *, out=None) -> Tensor`

Function signatures followed the NumPy-like style for the moment, keeping `out` at the end to be consistent with PyTorch.

### CODE REVIEW

**Comment by RockingJavaBean** [comment]
The screenshots of rendered docs for the `torch.quantile` function are below.
- [description](https://user-images.githubusercontent.com/40605299/101020255-2adffe80-35a9-11eb-8aea-c9022bb9b809.png)
- [example](https://user-images.githubusercontent.com/40605299/101020515-94600d00-35a9-11eb-984f-06b347853544.png)

**Comment by RockingJavaBean** [comment]
@mruberry  This PR implements additional quantile modes: `lower`, `higher`, `nearest`, and `midpoint` according to the function request.
Please kindly help review it.

**Comment by mruberry** [comment]
> @mruberry This PR implements additional quantile modes: `lower`, `higher`, `nearest`, and `midpoint` according to the function request.

Awesome!

> Please kindly help review it.

@heitorschueroff, who implemented PyTorch's quantile function, will help with the review.

File: aten/src/ATen/native/Sorting.cpp
Comment (heitorschueroff) [inline aten/src/ATen/native/Sorting.cpp:438]: This will break backwards compatibility for the C++ API. @mruberry Is there a better solution to add this parameter than adding 8 more overloads to native_functions?

File: aten/src/ATen/native/native_functions.yaml
Comment (heitorschueroff) [inline aten/src/ATen/native/native_functions.yaml:6731]: @mruberry Do we have a preference for whether dim and keepdim should be next to each other?

File: torch/_torch_docs.py
Comment (heitorschueroff) [inline torch/_torch_docs.py:5122]: ... based on ... -> . When the desired quantile lies between two data points \`\`i < j``, the result is computed based on the :attr:\`interpolation\` value as described below. By default, :attr:\`interpolation\` is \`"linear"\` and :attr:\`dim\` ...

File: torch/_torch_docs.py
Comment (heitorschueroff) [inline torch/_torch_docs.py:5125]: Describe this in terms of data points i and j similar to how NumPy does it. Something like:

When the quantile value lies between two data points \`\`i < j\`\`, the result is computed according to the given :attr:\`interpolation\` method as follows:

Then describe each method in terms of i and j.

File: torch/_torch_docs.py
Comment (heitorschueroff) [inline torch/_torch_docs.py:5144]: Add "Default is ``linear``" at the end.

File: torch/_torch_docs.py
Comment (heitorschueroff) [inline torch/_torch_docs.py:5170]: Great example. I would add the linear interpolation method and change the q value to 0.6 so there is a difference between linear and midpoint.

**Review by heitorschueroff — CHANGES_REQUESTED** [review]
Thank you @RockingJavaBean for the PR. It's heading on the right direction but there's some room for improving performance. I also left some comments on the docs and I need to check with @mruberry on the BC breaking change.

File: aten/src/ATen/native/native_functions.yaml
Comment (mruberry) [reply aten/src/ATen/native/native_functions.yaml:6731]: It would make conceptual sense if they were next to each other. Let's implement a NumPy-like signature for the moment (keeping out at the end to be consistent with PyTorch). In the future I think we should consider making these args kwarg-only.

File: aten/src/ATen/native/Sorting.cpp
Comment (mruberry) [reply aten/src/ATen/native/Sorting.cpp:438]: Good question. Probably not, unfortunately. Just mark the PR as "bc-breaking" with a label and add a "BC-breaking summary" at the top of the PR summary describing what's changed and why it's BC-breaking.

File: aten/src/ATen/native/Sorting.cpp
Comment (RockingJavaBean) [reply aten/src/ATen/native/Sorting.cpp]: Thanks for pointing this out, it has been moved next to the other checks.

File: aten/src/ATen/native/Sorting.cpp
Comment (RockingJavaBean) [reply aten/src/ATen/native/Sorting.cpp]: Thank you so much for providing valuable insights regarding performance optimization, and I have updated the code accordingly.

File: torch/_torch_docs.py
Comment (RockingJavaBean) [reply torch/_torch_docs.py]: Addressed.

File: torch/_torch_docs.py
Comment (RockingJavaBean) [reply torch/_torch_docs.py]: Thanks so much for the tip for improving the docs of `torch.quantile` function, I have pdated the docs based on the tip.

File: torch/_torch_docs.py
Comment (RockingJavaBean) [reply torch/_torch_docs.py]: Addressed.

File: torch/_torch_docs.py
Comment (RockingJavaBean) [reply torch/_torch_docs.py]: Thank you so much for pointing this out.
The linear interpolation method is added to the example, and the `q` value is changed to 0.6 to demonstrate the difference between `linear` and `midpoint`.

File: aten/src/ATen/native/native_functions.yaml
Comment (RockingJavaBean) [reply aten/src/ATen/native/native_functions.yaml:6731]: The NumPy's signature of quantile function is below:
`numpy.quantile(a, q, axis=None, out=None, overwrite_input=False, interpolation='linear', keepdims=False)`.

The current PR follows the NumPy style, by putting `interpolation` in front of `keepdims`.

I am a little bit uncertain about whether we should make `dim` and `keepdim` next to each other for the moment
Please kindly let me know if the order of `interpolation` and `keepdims` parameters should be exchanged, I will update this PR accordingly. @mruberry @heitorschueroff

File: aten/src/ATen/native/Sorting.cpp
Comment (RockingJavaBean) [reply aten/src/ATen/native/Sorting.cpp:445]: I updated the PR summary with bc-breaking notes for adding interpolation parameter to torch.quantile function, please kindly confirm it and may update the content.

**Comment by RockingJavaBean** [comment]
@mruberry @heitorschueroff I'm truly grateful for the kind help in reviewing this PR.
I have modified the code implementation based on the suggestions of performance optimization, the docs descriptions and examples are updated as well.
The `clang-tidy` error seems were caused by the failure of pulling the third-party code repository, and it seems is not related to this pull request.
Please kindly help review the updated PR.

**Comment by RockingJavaBean** [comment]
@heitorschueroff Gentle ping.
I have updated this PR according to the feedbacks of the code review and rebased the code, All CI checks seem passed.
Please kindly take a look.

File: aten/src/ATen/native/native_functions.yaml
Comment (heitorschueroff) [reply aten/src/ATen/native/native_functions.yaml:6728]: Let's keep it this way then for the moment.

**Review by heitorschueroff — APPROVED** [review]
LGTM. Thank you for this great contribution.

**Comment by matthewfeickert** [comment]
Thanks very much for this PR. This is great to have in!

**Comment by RockingJavaBean** [comment]
@heitorschueroff could you please share the cause of reverting this PR, is there anything I could do for help?

**Comment by heitorschueroff** [comment]
> @heitorschueroff could you please share the cause of reverting this PR, is there anything I could do for help?

@RockingJavaBean I reverted it because some internal tests were failing and it seems this PR was to blame. I'm investigating what could be the cause and I'll let you know when I have more information.

**Comment by heitorschueroff** [comment]
@RockingJavaBean This PR actually breaks BC compatibility of the python API as well. Before this PR, one could call `torch.quantile(samples, quantiles, 0, False)` but now this will not run because of the missing interpolation parameter. One solution is to move interpolation after keepdim but this would not be very standard with our API, @mruberry what do you think?

**Comment by mruberry** [comment]
I think we can support current users (with a deprecation warning) and the new API by adding signatures. It's not the prettiest solution but I think we'll have to do it. It's too bad the bc-break didn't appear in any of our pre-land tests. I'm surprised it happened, but I guess that shows how popular quantile is.

**Comment by heitorschueroff** [comment]
@RockingJavaBean I created a new PR here https://github.com/pytorch/pytorch/pull/49267 to fix the BC breaking issue. This was my fault for missing it in the review. At this point you don't need to take any further action, I'll land the PR once approved. Thanks for the contribution.

**Comment by RockingJavaBean** [comment]
@heitorschueroff Thank you so much for all the kind help and the insightful suggestions regarding improving performances, I'm truly thankful.

**Pull request closed without merging**
